// web3-plugin-contracts/lib/chains/chain-id.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChainID = void 0;
var ChainID;
(function (ChainID) {
    ChainID["ETHEREUM_MAINNET"] = "1";
    ChainID["ETHEREUM_GOERLI"] = "5";
    ChainID["ETHEREUM_SEPOLIA"] = "11155111";
    ChainID["ETHEREUM_HOLESKY"] = "17000";
    ChainID["METIS_ANDROMEDA"] = "1088";
    ChainID["METIS_SEPOLIA"] = "59902";
    ChainID["ARBITRUM_ONE"] = "42161";
    ChainID["ARBITRUM_GOERLI"] = "421613";
    ChainID["ARBITRUM_NOVA"] = "42170";
    ChainID["OASIS_EMERALD"] = "42262";
    ChainID["OASIS_SAPPHIRE"] = "23294";
    ChainID["OASIS_SAPPHIRE_TESTNET"] = "23295";
    ChainID["IMMUTABLE_MAINNET"] = "13371";
    ChainID["IMMUTABLE_TESTNET"] = "13473";
    ChainID["AVALANCHE_MAINNET"] = "43114";
    ChainID["AVALANCHE_FUJI"] = "43113";
    ChainID["POLYGON_MAINNET"] = "137";
    ChainID["POLYGON_MUMBAI"] = "80001";
    ChainID["OPTIMISM_MAINNET"] = "10";
    ChainID["OPTIMISM_SEPOLIA"] = "11155420";
    ChainID["FANTOM_MAINNET"] = "250";
    ChainID["FANTOM_TESTNET"] = "4002";
    ChainID["BASE_MAINNET"] = "8453";
    ChainID["BASE_SEPOLIA"] = "84532";
    ChainID["BNB_MAINNET"] = "56";
    ChainID["BNB_TESTNET"] = "97";
    ChainID["LINEA_MAINNET"] = "59144";
    ChainID["LINEA_TESTNET"] = "59140";
    ChainID["MOONBEAM_MAINNET"] = "1284";
    ChainID["MOONBASE_ALPHA"] = "1287";
    ChainID["MOONRIVER_MAINNET"] = "1285";
    ChainID["CELO_MAINNET"] = "42220";
    ChainID["CELO_ALFAJORES"] = "44787";
    ChainID["GNOSIS_MAINNET"] = "100";
    ChainID["CRONOS_MAINNET"] = "25";
    ChainID["SCROLL_MAINNET"] = "534351";
    ChainID["SCROLL_SEPOLIA"] = "534352";
    ChainID["MANTLE_MAINNET"] = "5000";
    ChainID["MANTLE_TESTNET"] = "5001";
    ChainID["BITTORRENT_MAINNET"] = "199";
    ChainID["BITTORRENT_TESTNET"] = "1029";
    ChainID["XDC_MAINNET"] = "50";
    ChainID["CANTO_MAINNET"] = "7700";
    ChainID["CANTO_TESTNET"] = "7701";
    ChainID["KAVA_MAINNET"] = "2222";
    ChainID["KAVA_TESTNET"] = "2221";
    ChainID["ROLLUX_MAINNET"] = "570";
    ChainID["ROLLUX_TESTNET"] = "57000";
    ChainID["SYSCOIN_MAINNET"] = "57";
    ChainID["SYSCOIN_TESTNET"] = "5700";
    ChainID["BOBA_ETHEREUM"] = "288";
    ChainID["MANTA_PACIFIC"] = "169";
    ChainID["MANTA_TESTNET"] = "3441005";
    ChainID["WEMIX_MAINNET"] = "1111";
    ChainID["WEMIX_TESTNET"] = "1112";
    ChainID["ASTAR_MAINNET"] = "592";
    ChainID["SHIDEN_MAINNET"] = "336";
    ChainID["SHUBIYA_TESTNET"] = "81";
    ChainID["ZETACHAIN_MAINNET"] = "7000";
    ChainID["ZETACHAIN_TESTNET"] = "7001";
    ChainID["FLARE_MAINNET"] = "14";
    ChainID["FLARE_COSTON"] = "16";
    ChainID["FLARE_COSTON2"] = "114";
    ChainID["FUSE_MAINNET"] = "122";
    ChainID["FUSE_SPARK"] = "123";
    ChainID["SHIBARIUM_MAINNET"] = "156";
    ChainID["PUPPYNET_TESTNET"] = "157";
    ChainID["CONFLUX_MAINNET"] = "1030";
    ChainID["CONFLUX_TESTNET"] = "71";
    ChainID["FILECOIN_MAINNET"] = "314";
    ChainID["FILECOIN_CALIBRATION"] = "314159";
    ChainID["ENERGY_WEB_CHAIN"] = "246";
    ChainID["ENERGY_WEB_CHAIN_TESTNET"] = "73799";
    ChainID["LUKSO_MAINNET"] = "42";
    ChainID["LUKSO_TESTNET"] = "4201";
    ChainID["ACALA_MAINNET"] = "787";
    ChainID["MANDALA_TESTNET"] = "595";
    ChainID["REI_MAINNET"] = "47805";
    ChainID["REI_TESTNET"] = "12357";
    ChainID["ZORA_NETWORK_MAINNET"] = "7777777";
    ChainID["ZORA_NETWORK_TESTNET"] = "999999999";
    ChainID["CALLISTO_MAINNET"] = "820";
    ChainID["ZKSYNC_MAINNET"] = "324";
    ChainID["ZKSYNC_SEPOLIA"] = "300";
    ChainID["NEON_MAINNET"] = "245022934";
    ChainID["NEON_TESTNET"] = "245022935";
    ChainID["AURORA_MAINNET"] = "1313161554";
    ChainID["AURORA_TESTNET"] = "1313161555";
    ChainID["RONIN_MAINNET"] = "2020";
    ChainID["RONIN_SAIGON_TESTNET"] = "2021";
    ChainID["TRON_MAINNET"] = "tron";
    ChainID["TRON_NILE_TESTNET"] = "tron_nile";
    ChainID["TRON_SHASTA_TESTNET"] = "tron_shasta";
    ChainID["BEAM_MAINNET"] = "4337";
    ChainID["CARBON_EVM_MAINNET"] = "9790";
    ChainID["CARBON_EVM_TESTNET"] = "9792";
    ChainID["ARTELA_TESTNET"] = "11822";
    ChainID["BERACHAIN_TESTNET"] = "80085";
    ChainID["VICTION_MAINNET"] = "88";
    ChainID["VICTION_TESTNET"] = "89";
    ChainID["PALM_MAINNET"] = "11297108109";
    ChainID["PALM_TESTNET"] = "11297108099";
    ChainID["METER_MAINNET"] = "82";
    ChainID["METER_TESTNET"] = "83";
    ChainID["PUBLIC_GOOD_NETWORK"] = "424";
    ChainID["PUBLIC_GOOD_NETWORK_SEPOLIA"] = "58008";
    ChainID["ROOTSTOCK_MAINNET"] = "30";
    ChainID["ROOTSTOCK_TESTNET"] = "31";
    ChainID["LIGHTLINK_PHOENIX_MAINNET"] = "1890";
    ChainID["LIGHTLINK_PEGASUS_TESTNET"] = "1891";
    ChainID["ETHERLINK_TESTNET"] = "128123";
    ChainID["SHARDEUM_SPHINX_1_X"] = "8082";
    ChainID["VELAS_MAINNET"] = "106";
    ChainID["MODE_MAINNET"] = "34443";
    ChainID["MODE_SEPOLIA"] = "919";
    ChainID["MORPH_TESTNET"] = "2710";
    ChainID["BLAST_MAINNET"] = "81457";
    ChainID["BLAST_SEPOLIA"] = "168587773";
    ChainID["SYNDICATE_FRAME_MAINNET"] = "5101";
    ChainID["DOS_MAINNET"] = "7979";
    ChainID["DOS_TESTNET"] = "3939";
    ChainID["DEGEN_MAINNET"] = "666666666";
    ChainID["TAIKO_KATLA_TESTNET"] = "167008";
    ChainID["SHIMMER_MAINNET"] = "148";
    ChainID["SHIMMER_TESTNET"] = "1073";
    ChainID["FRAXSCAN_MAINNET"] = "252";
    ChainID["FRAXSCAN_TESTNET"] = "2522";
    ChainID["XRP_SIDECHAIN"] = "1440002";
    ChainID["CORE_MAINNET"] = "1116";
    ChainID["ZKEVM_POLYGON"] = "1101";
    ChainID["ZKEVM_POLYGON_CARDONA"] = "2442";
    ChainID["BITLAYER_MAINNET"] = "200901";
    ChainID["BITLAYER_TESTNET"] = "200810";
    ChainID["ETHEREUM_CLASSIC_MAINNET"] = "61";
    ChainID["ETHEREUM_CLASSIC_TESTNET"] = "62";
    ChainID["STABILITY_MAINNET"] = "101010";
    ChainID["STABILITY_TESTNET"] = "20180427";
    ChainID["ARBITRUM_SEPOLIA"] = "421614";
    ChainID["KROMA_MAINNET"] = "255";
    ChainID["KROMA_SEPOLIA"] = "2358";
    ChainID["POLYGON_AMOY"] = "80002";
    ChainID["GNOSIS_CHIADO"] = "10200";
    ChainID["COTI_DEVNET"] = "13068200";
    ChainID["LISK_MAINNET"] = "1135";
    ChainID["LISK_SEPOLIA"] = "4202";
    ChainID["REDSTONE_MAINNET"] = "690";
    ChainID["REDSTONE_GARNET_TESTNET"] = "17069";
    ChainID["OPEN_CAMPUS_CODEX"] = "656476";
})(ChainID || (exports.ChainID = ChainID = {}));
