// web3-plugin-contracts/lib/chains/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTronRPC = exports.getRPC = exports.getNetworkNameFromChainID = exports.getAPIKey = exports.getIconByChainId = exports.getContractExplorer = exports.getExplorer = exports.ChainID = exports.getAPI = void 0;
var api_1 = require("./api");
Object.defineProperty(exports, "getAPI", { enumerable: true, get: function () { return api_1.getAPI; } });
var chain_id_1 = require("./chain-id");
Object.defineProperty(exports, "ChainID", { enumerable: true, get: function () { return chain_id_1.ChainID; } });
var explorer_1 = require("./explorer");
Object.defineProperty(exports, "getExplorer", { enumerable: true, get: function () { return explorer_1.getExplorer; } });
Object.defineProperty(exports, "getContractExplorer", { enumerable: true, get: function () { return explorer_1.getContractExplorer; } });
var icon_1 = require("./icon");
Object.defineProperty(exports, "getIconByChainId", { enumerable: true, get: function () { return icon_1.getIconByChainId; } });
var key_1 = require("./key");
Object.defineProperty(exports, "getAPIKey", { enumerable: true, get: function () { return key_1.getAPIKey; } });
var name_1 = require("./name");
Object.defineProperty(exports, "getNetworkNameFromChainID", { enumerable: true, get: function () { return name_1.getNetworkNameFromChainID; } });
var rpc_1 = require("./rpc");
Object.defineProperty(exports, "getRPC", { enumerable: true, get: function () { return rpc_1.getRPC; } });
var tron_1 = require("./tron");
Object.defineProperty(exports, "getTronRPC", { enumerable: true, get: function () { return tron_1.getTronRPC; } });
