// web3-plugin-contracts/lib/solidity/metadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.clean = exports.settings = exports.compilerVersion = exports.contractName = exports.abi = exports.language = void 0;
const language = (metadata) => {
    return metadata?.language || "Solidity";
};
exports.language = language;
const abi = (metadata) => {
    return JSON.stringify(metadata?.output?.abi || {});
};
exports.abi = abi;
const contractName = (metadata) => {
    if (metadata?.settings?.compilationTarget) {
        const key = Object.keys(metadata.settings.compilationTarget).pop() || "";
        if (key) {
            return metadata.settings.compilationTarget[key];
        }
    }
    return "";
};
exports.contractName = contractName;
const compilerVersion = (metadata) => {
    if (metadata?.compiler?.version) {
        return metadata.compiler.version;
    }
    return "";
};
exports.compilerVersion = compilerVersion;
const settings = (metadata) => {
    if (!metadata?.settings) {
        return {};
    }
    if (metadata?.settings?.libraries) {
        delete metadata.settings.libraries;
    }
    if (metadata?.settings?.compilationTarget) {
        delete metadata.settings.compilationTarget;
    }
    if (metadata?.settings?.remappings) {
        delete metadata.settings.remappings;
    }
    if (metadata?.settings?.metadata) {
        delete metadata.settings.metadata;
    }
    return metadata.settings;
};
exports.settings = settings;
const clean = (metadata) => {
    metadata.settings = (0, exports.settings)(metadata);
    return metadata;
};
exports.clean = clean;
